packages/app/src/heartbeatService.ts

```diff
@@ -0,0 +1,161 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ComponentContainer } from '@firebase/component';
+import {
+  base64Encode,
+  isIndexedDBAvailable,
+  validateIndexedDBOpenable
+} from '@firebase/util';
+import {
+  deleteHeartbeatsFromIndexedDB,
+  readHeartbeatsFromIndexedDB,
+  writeHeartbeatsToIndexedDB
+} from './indexeddb';
+import { FirebaseApp } from './public-types';
+import {
+  HeartbeatsByUserAgent,
+  HeartbeatService,
+  HeartbeatStorage
+} from './types';
+
+export class HeartbeatServiceImpl implements HeartbeatService {
+  storage: HeartbeatStorageImpl;
+  heartbeatsCache: HeartbeatsByUserAgent[] | null = null;
+  heartbeatsCachePromise: Promise<HeartbeatsByUserAgent[]>;
+  constructor(private readonly container: ComponentContainer) {
+    const app = this.container.getProvider('app').getImmediate();
+    this.storage = new HeartbeatStorageImpl(app);
+    this.heartbeatsCachePromise = this.storage
+      .read()
+      .then(result => (this.heartbeatsCache = result));
+  }
+  async triggerHeartbeat(): Promise<void> {
+    const platformLogger = this.container
+      .getProvider('platform-logger')
+      .getImmediate();
+    const userAgent = platformLogger.getPlatformInfoString();
+    const date = getDateString();
+    if (!this.heartbeatsCache) {
+      await this.heartbeatsCachePromise;
+    }
+    let heartbeatsEntry = this.heartbeatsCache!.find(
+      heartbeats => heartbeats.userAgent === userAgent
+    );
+    if (heartbeatsEntry) {
+      if (heartbeatsEntry.dates.includes(date)) {
+        return;
+      } else {
+        heartbeatsEntry.dates.push(date);
+      }
+    } else {
+      heartbeatsEntry = {
+        userAgent,
+        dates: [date]
+      };
+    }
+    return this.storage.overwrite([]);
+  }
+  async getHeartbeatsHeader(): Promise<string> {
+    if (!this.heartbeatsCache) {
+      await this.heartbeatsCachePromise;
+    }
+    return base64Encode(JSON.stringify(this.heartbeatsCache!));
+  }
+}
+
+function getDateString(): string {
+  const today = new Date();
```

Comment:
We need to send the `UTC` date, so you should use the UTC version of the methods, e.g. `getUTCFullYear()`.